File: src/210-practice.ts
const findAllOrdersIterative = (numCourses: number, prerequisites: number[][]) => {
  const graph: number[][] = Array.from({ length: numCourses }, () => [])
  const initialIndegree = Array(numCourses).fill(0)
  const initialUsed = Array.from({ length: numCourses }, () => false)

  type State = {
    path: number[],
    indegree: number[],
    used: boolean[]
  }

  const result: number[][] = []

  for (const [pre, cur] of prerequisites) {
    graph[cur].push(pre)
    initialIndegree[pre]++
  }

  const stack: State[] = [{
    path: [],
    indegree: initialIndegree,
    used: initialUsed
  }]

  while (stack.length > 0) {
    const { path, indegree, used } = stack.pop()!
    //exit
    if (path.length === numCourses) {
      result.push(path)
      continue
    }
    const candidates: number[] = []

    for (let course = 0; course < numCourses; course++) {
      if (!used[course] && indegree[course] === 0) {
        candidates.push(course)
      }
    }

    for (const candidate of candidates) {
      if (indegree[candidate] === 0 && !used[candidate]) {

        const newPath = [...path, candidate]
        const newIndegree = [...indegree]
        const newUser = [...used]

        newUser[candidate] = true;

        for (const next of graph[candidate]) {
          newIndegree[next]--;
        }

        stack.push({
          path: newPath,
          indegree: newIndegree,
          used: newUser
        })
      }
    }
  }

  return result
}

const findOneOrder = (numCourses: number, prerequisites: number[][]) => {
  const used: boolean[] = Array(numCourses).fill(false)
  const indegree: number[] = Array(numCourses).fill(0)
  const graph: number[][] = Array.from({ length: numCourses }, () => [])

  for (const [pre, cur] of prerequisites) {
    graph[cur].push(pre)
    indegree[pre]++
  }

  const path: number[] = []

  for (let course = 0; course < numCourses; course++) {
    if (indegree[course] === 0 && used[course] === false) {
      used[course] = true
      path.push(course)


      for (const next of graph[course]) {
        indegree[next]--
      }
    }
  }

  return path
}

const findAllOrderBacktracking = (numCourses: number, prerequest: number[][]) => {
  const used: boolean[] = Array(numCourses).fill(false)
  const indegree: number[] = Array(numCourses).fill(0)
  const graph: number[][] = Array.from({ length: numCourses }, () => [])

  for (const [cur, pre] of prerequest) {
    graph[pre].push(cur)
    indegree[cur]++
  }

  const result: number[][] = []
  const path: number[] = []
  const backtracking = () => {
    if (path.length === numCourses) {
      result.push([...path])
      return;
    }

    let candidate = false

    for (let course = 0; course < numCourses; course++) {
      if (!used[course] && indegree[course] === 0) {
        candidate = true
        used[course] = true

        path.push(course)
        for (const next of graph[course]) {
          indegree[next]--
        }

        backtracking()

        for (const next of graph[course]) {
          indegree[next]++
        }

        path.pop()
        used[course] = false
      }
    }

    if (!candidate) {
      return
    }
  }

  backtracking()
  return result

}


// ---- tests ----
{
  const normalize = (orders: number[][]): number[][] =>
    [...orders].sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)))

  const check = (name: string, actual: number[][], expected: number[][]): void => {
    const a = JSON.stringify(normalize(actual))
    const e = JSON.stringify(normalize(expected))
    console.log(a === e ? `PASS ${name}` : `FAIL ${name}: expected ${e}, got ${a}`)
  }

  type Case = { name: string, numCourses: number, prerequisites: number[][], expected: number[][] }
  const cases: Case[] = [
    {
      name: "case1 diamond",
      numCourses: 4,
      prerequisites: [[1, 0], [2, 0], [3, 1], [3, 2]],
      expected: [[0, 1, 2, 3], [0, 2, 1, 3]],
    },
    {
      name: "case2 simple chain",
      numCourses: 2,
      prerequisites: [[1, 0]],
      expected: [[0, 1]],
    },
    {
      name: "case3 cycle has no order",
      numCourses: 2,
      prerequisites: [[1, 0], [0, 1]],
      expected: [],
    },
  ]

  for (const c of cases) {
    check(`${c.name} (backtracking)`, findAllOrderBacktracking(c.numCourses, c.prerequisites), c.expected)
    check(`${c.name} (iterative)`, findAllOrdersIterative(c.numCourses, c.prerequisites), c.expected)
  }

  // findOneOrder is a single forward pass, so it only works when courses
  // unlock in increasing index order — test it on such a case only
  check("case4 findOneOrder diamond", [findOneOrder(4, [[1, 0], [2, 0], [3, 1], [3, 2]])], [[0, 1, 2, 3]])
}

// make this file a module so its declarations stay file-scoped
export {}

File: src_py/210-practice.py
"""
210. Course Schedule II - practice variants
Difficulty: Medium

Extra practice implementations: one valid order, all orders via backtracking,
and all orders via an explicit stack (iterative).
"""

from typing import List


def find_one_order(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    """Kahn's algorithm producing a single topological order (or []). """
    graph: List[List[int]] = [[] for _ in range(num_courses)]
    indegree = [0] * num_courses
    for course, prereq in prerequisites:
        graph[prereq].append(course)
        indegree[course] += 1

    stack = [c for c in range(num_courses) if indegree[c] == 0]
    order: List[int] = []
    while stack:
        node = stack.pop()
        order.append(node)
        for nxt in graph[node]:
            indegree[nxt] -= 1
            if indegree[nxt] == 0:
                stack.append(nxt)

    return order if len(order) == num_courses else []


def find_all_orders_backtracking(num_courses: int, prerequisites: List[List[int]]) -> List[List[int]]:
    graph: List[List[int]] = [[] for _ in range(num_courses)]
    indegree = [0] * num_courses
    for course, prereq in prerequisites:
        graph[prereq].append(course)
        indegree[course] += 1

    result: List[List[int]] = []
    path: List[int] = []
    used = [False] * num_courses

    def backtrack() -> None:
        if len(path) == num_courses:
            result.append(path[:])
            return
        for course in range(num_courses):
            if not used[course] and indegree[course] == 0:
                used[course] = True
                path.append(course)
                for nxt in graph[course]:
                    indegree[nxt] -= 1
                backtrack()
                for nxt in graph[course]:
                    indegree[nxt] += 1
                path.pop()
                used[course] = False

    backtrack()
    return result


def find_all_orders_iterative(num_courses: int, prerequisites: List[List[int]]) -> List[List[int]]:
    graph: List[List[int]] = [[] for _ in range(num_courses)]
    initial_indegree = [0] * num_courses
    for course, prereq in prerequisites:
        graph[prereq].append(course)
        initial_indegree[course] += 1

    result: List[List[int]] = []
    # State: (path, indegree, used)
    stack = [([], initial_indegree, [False] * num_courses)]

    while stack:
        path, indegree, used = stack.pop()
        if len(path) == num_courses:
            result.append(path)
            continue
        for course in range(num_courses):
            if not used[course] and indegree[course] == 0:
                new_indegree = indegree[:]
                new_used = used[:]
                new_used[course] = True
                for nxt in graph[course]:
                    new_indegree[nxt] -= 1
                stack.append((path + [course], new_indegree, new_used))

    return result


if __name__ == "__main__":
    prereq = [[1, 0], [2, 0], [3, 1], [3, 2]]
    expected = [[0, 1, 2, 3], [0, 2, 1, 3]]

    one = find_one_order(4, prereq)
    assert one[0] == 0 and one[-1] == 3

    assert sorted(find_all_orders_backtracking(4, prereq)) == sorted(expected)
    assert sorted(find_all_orders_iterative(4, prereq)) == sorted(expected)
    print("210-practice OK:", find_all_orders_iterative(4, prereq))
